presto-main/src/main/java/com/facebook/presto/split/BufferingSplitSource.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.split;
+
+import com.facebook.presto.connector.ConnectorId;
+import com.facebook.presto.metadata.Split;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class BufferingSplitSource
+        implements SplitSource
+{
+    private final int bufferSize;
+    private final SplitSource source;
+
+    public BufferingSplitSource(SplitSource source, int bufferSize)
+    {
+        this.source = requireNonNull(source, "source is null");
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public ConnectorId getConnectorId()
+    {
+        return source.getConnectorId();
+    }
+
+    @Override
+    public CompletableFuture<List<Split>> getNextBatch(int maxSize)
+    {
+        checkArgument(maxSize > 0, "Cannot fetch a batch of zero size");
+        ConcurrentLinkedQueue<Split> result = new ConcurrentLinkedQueue<>();
```

Comment:
ConcurrentLinkedQueue is pretty expensive.  I'd use `Collections.synchronizedList(new ArrayList<>(maxSize))` instead (with static import).